### ISSUE TITLE
Two forward passes during training

### ISSUE THREAD

**Issue by rometaidla** [PrieureDeSion/drive-any-robot#20]
What is the reason of doing two forward passed through the model during training to calculate losses for distance and action? Is it just implementation detail or there is some reason to do it this way related to propagating loss backward?

https://github.com/PrieureDeSion/drive-any-robot/blob/b58b2d75d4153f28f3c0ce9a68d291822b9cf263/train/gnm_train/training/train_utils.py#L239
https://github.com/PrieureDeSion/drive-any-robot/blob/b58b2d75d4153f28f3c0ce9a68d291822b9cf263/train/gnm_train/training/train_utils.py#L246

**Comment by rometaidla** [comment]
Understood from the code. Goal sampling parameters are different for distance and action.

**Comment by rometaidla** [comment]
I reopen as I don't would like to understand this better. As I commented previously seems like the two separate forward passes are done separately as distance and goal sampling seems to have different hyper-parameters, so goal image is passed throught the network is different.

What is the reason to have different parameters for goal sampling? Wouldn't using same goal for both distance and action prediction (and doing only one forward pass) give similar result?

**Comment by ajaysridhar0** [comment]
Hi @rometaidla,

Yeah, this is an interesting design choice in our training pipeline. We use different goal-sampling strategies for learning the distance and action output because of how we deploy the model. We only need to predict actions for subgoals close to the robot since faraway subgoals would be out of sight from the robot. However, we need to be able to determine if a goal is distant with the distance head, so we increase the range of goals for distance learning.

As for the two forward passes, we fixed this inefficiency in our new repo [visualnav-transformer](https://github.com/robodhruv/visualnav-transformer/tree/main) by masking the action loss function for distant goals. This repo also contains follow ups to our work on general navigation models if you are curious.

--Ajay
